### PR TITLE
fix(java): initialize classes at build-time to fix Native Image compilation

### PR DESCRIPTION
**Pull request by mpeddada1**
Using META-INF/services with Native Image compilation causes the `FileSystemProvider` class to be initialized at build time. This causes `com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider` and a couple of unexpected classes that are referenced to also be initialized at build time.

### CODE REVIEW

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Using META-INF/services with Native Image compilation results in
+# FileSystemProvider being initialized at build time. This results
+# CloudStorageFileSystemProvider and all the sub-classes referenced by
+# this class also being initialized at build time.
```
Comment (suztomo) [inline google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties:4]: Only for static fields or static initializers are needed.

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+Args = --initialize-at-build-time=com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider,\
+  com.google.cloud.storage.contrib.nio.CloudStorageConfiguration,\
+  com.google.cloud.storage.contrib.nio.CloudStorageFileSystem,\
+  com.google.cloud.storage.contrib.nio.StorageOptionsUtil,\
+  com.google.cloud.storage.StorageException,\
+  com.google.cloud.storage.StorageOptions,\
+  com.google.cloud.storage.DefaultStorageRetryStrategy,\
+  com.google.api.client.http.OpenCensusUtils,\
+  com.google.api.client.http.GenericUrl,\
```
Comment (suztomo) [inline google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties:13]: (I'm curious about how you get the list)

If we don't have this `GenericUrl` line here, what kind of error do you get?

**Review by meltsufin — APPROVED** [review]

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties (hunk 2 above)
Comment (mpeddada1) [reply google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties:13]: Removing `GenericUrl` results in:
```
Error: Classes that should be initialized at run time got initialized during image building:
 com.google.api.client.http.GenericUrl was unintentionally initialized at build time. To see why com.google.api.client.http.GenericUrl got initialized use --trace-class-initialization=com.google.api.client.http.GenericUrl
```

And removing all the other classes results in a similar for the following classes:
 
```
--trace-class-initialization=com.google.api.client.json.gson.GsonParser$1,io.opencensus.trace.Tracing,io.opencensus.trace.EndSpanOptions,com.google.api.client.http.HttpRequest,org.threeten.bp.Duration,com.google.api.client.http.javanet.NetHttpTransport,com.google.auth.oauth2.OAuth2Utils,com.google.cloud.storage.StorageOptions,io.opencensus.contrib.http.util.CloudTraceFormat,com.google.auth.oauth2.DefaultCredentialsProvider,io.opencensus.trace.Status,io.opencensus.trace.AutoValue_EndSpanOptions,com.google.cloud.storage.DefaultStorageRetryStrategy,com.google.api.client.http.OpenCensusUtils,com.google.cloud.storage.StorageException,com.google.gson.stream.JsonReader,com.google.auth.oauth2.OAuth2Credentials,com.google.api.client.util.DateTime,com.google.api.client.util.escape.PercentEscaper,com.google.cloud.ExceptionHandler,io.opencensus.trace.BlankSpan,io.opencensus.trace.SpanContext,com.google.api.client.json.JsonParser$1,com.google.api.client.http.HttpTransport,com.google.api.client.http.HttpMediaType,com.google.auth.oauth2.GoogleCredentials,com.google.common.collect.RegularImmutableMap,io.opencensus.trace.Span,com.google.api.client.http.GenericUrl,com.google.cloud.ServiceOptions,com.google.api.client.util.Data,com.google.api.client.util.escape.CharEscapers,com.google.cloud.PlatformInformation
```

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties (hunk 1 above)
Comment (mpeddada1) [reply google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties]: Slightly updated the comment here. Let me know if this sounds okay to you? Explored the code a little further to double check what's happening: The initialization of classes stems from `CloudStorageFileSystemProvider` which causes `StorageOptions` and as a result `StorageOptionsUtil`  to be initialized at build time.
https://github.com/googleapis/java-storage-nio/blob/004ab8c95dad31958701851381267e00e207da6a/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java#L96

Looks like this seems to be the case in the [reproducer](https://github.com/mpeddada1/junit-experiment) as well where when we create a new class `MyOptions` and instantiate it in `MySampleFileSystemProvider`, `MyOptions` ends up getting initialized at build time.

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties (hunk 2 above)
Comment (suztomo) [reply google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties:15]: I see. Thank you for checking.

**Review by suztomo — APPROVED** [review]

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  com.google.common.cache.LocalCache,\
+  com.google.common.cache.CacheBuilder,\
+  com.google.common.math.IntMath$1,\
+  com.google.common.collect.RegularImmutableMap,\
+  com.google.gson.stream.JsonReader,\
+  com.google.api.client.json.gson.GsonParser$1,\
+  com.google.api.client.json.JsonParser$1,\
+  com.google.cloud.PlatformInformation,\
+  com.google.cloud.ExceptionHandler,\
+  org.threeten.bp.Duration
```
Comment (BenWhitehead) [inline google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties:38]: I had to add the following additional classes when running via
```bash
export JAVA_HOME=~/opt/java/11-native
mvn -Pnative clean verify
```

```bash
> $JAVA_HOME/bin/java -version
openjdk version "11.0.14" 2022-01-18
OpenJDK Runtime Environment GraalVM CE 22.0.0.2 (build 11.0.14+9-jvmci-22.0-b05)
OpenJDK 64-Bit Server VM GraalVM CE 22.0.0.2 (build 11.0.14+9-jvmci-22.0-b05, mixed mode, sharing)
```

```suggestion
  org.threeten.bp.Duration,\
  com.google.common.io.BaseEncoding$Base16Encoding,\
  com.google.common.io.BaseEncoding$Base64Encoding,\
  com.google.common.io.BaseEncoding,\
  com.google.common.io.BaseEncoding$StandardBaseEncoding,\
  com.google.api.client.util.PemReader

```

**Review by BenWhitehead — CHANGES_REQUESTED** [review]
I had to add some additional classes, not sure if that's unexpected

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties (hunk 3 above)
Comment (mpeddada1) [reply google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties]: Thank you for verifying! Added these classes in. I had to skip running few tests due to project setup related issues so may have missed these when I commented them out.

**Review by BenWhitehead — APPROVED** [review]

File: google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties (hunk 3 above)
Comment (BenWhitehead) [reply google-cloud-nio/src/main/resources/META-INF/native-image/com/google/cloud/google-cloud-nio/native-image.properties]: Ah okay, thanks for confirming

**Pull request merged**